### PR TITLE
Host Resovler Arch Update

### PR DESCRIPTION
**Pull request by waahm7**
Issue #143 

*Description of changes:*
- Updates host resolver architecture
- Adds recordConnectionOnFailure, purgeCache, purgeCacheAddress

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift [hunk 1]
```diff
@@ -4,11 +4,27 @@
...
+    /// Resolves the address(es) for HostResolverArguments and returns a list of
+    /// addresses with (most likely) two addresses, one AAAA and one A.
+    func resolveAddress(args: HostResolverArguments) async throws -> [HostAddress]
+    /// Reports a failure on an address so that the background cache can accommodate
+    /// the failure and likely not return the address until it recovers.
+    func reportFailureOnAddress(address: HostAddress) throws
+    /// Empties the cache for an address.
+    func purgeCache(args: HostResolverArguments) async throws
+    /// Empties the cache for all addresses.
+    func purgeCache() async throws
```
Comment (graebm) [inline Source/AwsCommonRuntimeKit/io/HostResolver.swift:18]: doesn't seem like these last 3 functions should be throwing.

like, how could the user meaningfully respond to any of these failures

Oh, the underlying C functions can report an error.
Maybe this is a good time to change that

Comment (waahm7) [reply Source/AwsCommonRuntimeKit/io/HostResolver.swift]: Thanks, I have updated the Swift API to ignore the error.

**Review by graebm — APPROVED** [review]

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift [hunk 2]
```diff
@@ -59,15 +78,58 @@ public class HostResolver {
...
+        let host = AWSString(args.hostName, allocator: allocator)
+        return await withCheckedContinuation({ (continuation: CheckedContinuation<(), Never>) in
+            let continuationCore = Box(continuation)
+            var purgeCacheOptions = aws_host_resolver_purge_host_options()
+            purgeCacheOptions.host = UnsafePointer(host.rawValue)
+            purgeCacheOptions.on_host_purge_complete_callback = onPurgeCacheComplete
+            purgeCacheOptions.user_data = continuationCore.passRetained()
+            guard aws_host_resolver_purge_host_cache(rawValue, &purgeCacheOptions) == AWS_OP_SUCCESS else {
+                continuationCore.release()
+                return
```
Comment (graebm) [inline Source/AwsCommonRuntimeKit/io/HostResolver.swift:111]: forgot to resume continuation on failure

Comment (waahm7) [reply Source/AwsCommonRuntimeKit/io/HostResolver.swift:112]: Good catch, fixed.

File: Source/AwsCommonRuntimeKit/io/HostAddressType.swift [hunk 3]
```diff
@@ -3,21 +3,22 @@
...
+    case AAAA
 }
 
-extension AddressRecordType: RawRepresentable, CaseIterable {
+extension HostAddressType: RawRepresentable, CaseIterable {
 
     public init(rawValue: aws_address_record_type) {
         let value = Self.allCases.first { $0.rawValue == rawValue }
-        self = value ?? .typeA
+        self = value ?? .A
```
Comment (ganeshnj) [inline Source/AwsCommonRuntimeKit/io/HostAddressType.swift:16]: not related to this PR but we should make these type of init throwable.

Falling back on .A is bad for future work, for example if we have IPv8.

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift [hunk 4]
```diff
@@ -4,11 +4,27 @@
 import AwsCCommon
 import AwsCIo
 
-public class HostResolver {
+/// All host resolver implementation must conform to this protocol.
+public protocol HostResolverProtocol {
+    /// Resolves the address(es) for HostResolverArguments and returns a list of
+    /// addresses with (most likely) two addresses, one AAAA and one A.
+    func resolveAddress(args: HostResolverArguments) async throws -> [HostAddress]
```
Comment (ganeshnj) [inline Source/AwsCommonRuntimeKit/io/HostResolver.swift:11]: use Swift Doc than just plain ///

https://github.com/apple/swift/blob/main/docs/DocumentationComments.md

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift [hunk 5]
```diff
@@ -59,15 +78,60 @@ public class HostResolver {
...
+            let continuationCore = Box(continuation)
+            var purgeCacheOptions = aws_host_resolver_purge_host_options()
+            purgeCacheOptions.host = UnsafePointer(host.rawValue)
+            purgeCacheOptions.on_host_purge_complete_callback = onPurgeCacheComplete
+            purgeCacheOptions.user_data = continuationCore.passRetained()
+            guard aws_host_resolver_purge_host_cache(rawValue, &purgeCacheOptions) == AWS_OP_SUCCESS else {
+                continuationCore.release()
+                continuation.resume()
+                return
+            }
```
Comment (ganeshnj) [inline Source/AwsCommonRuntimeKit/io/HostResolver.swift:113]: if guard's else is not executed, we are not resuming continuation. There must be at least one resume.

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift [hunk 6]
```diff
@@ -59,15 +78,60 @@ public class HostResolver {
...
+            guard aws_host_resolver_purge_cache_with_callback(
+                    rawValue,
+                    onPurgeCacheComplete,
+                    continuationCore.passRetained()) == AWS_OP_SUCCESS
+            else {
+                continuationCore.release()
+                continuation.resume()
+                return
+            }
+        })
```
Comment (ganeshnj) [inline Source/AwsCommonRuntimeKit/io/HostResolver.swift:132]: Shouldn't there be a resume as well?

**Review by ganeshnj — CHANGES_REQUESTED** [review]

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift (hunk 5 above)
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/io/HostResolver.swift:113]: It is in the callback.

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift (hunk 6 above)
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/io/HostResolver.swift:132]: It is in the callback.

File: Source/AwsCommonRuntimeKit/io/HostAddressType.swift (hunk 3 above)
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/io/HostAddressType.swift:16]: Yeah, we should refactor all enums to a better pattern.

File: Source/AwsCommonRuntimeKit/io/HostResolver.swift (hunk 4 above)
Comment (waahm7) [reply Source/AwsCommonRuntimeKit/io/HostResolver.swift:14]: Thanks, fixed.

**Review by ganeshnj — APPROVED** [review]

**Pull request merged**